### PR TITLE
Add support for watching multiple entities at once

### PR DESCRIPTION
**Pull request by simon3z**

### CODE REVIEW

File: lib/kubeclient.rb [hunk 1]
```diff
@@ -18,9 +18,9 @@ class Client
     attr_reader :api_endpoint
     ENTITIES = %w(Pod Service ReplicationController Node Event)
 
-    def initialize(api_endpoint, version)
-      api_endpoint += '/' unless api_endpoint.end_with? '/'
-      @api_endpoint = api_endpoint + version
+    def initialize(uri, version)
```
Comment (abonas) [inline lib/kubeclient.rb]: why isn't this a separate PR?

**Comment by simon3z** [comment]
I'll do this by incremental steps.

File: lib/kubeclient.rb [hunk 2]
```diff
@@ -138,5 +129,51 @@ def all_entities
         result_hash[key_name] = send(method_name)
       end
     end
+
+    def watch(entities)
+      watch_endpoint = api_endpoint + '/watch/'
```
Comment (Fryguy) [inline lib/kubeclient.rb]: Prefer using URI, if you can...

File: lib/kubeclient.rb [hunk 3]
```diff
@@ -138,5 +129,51 @@ def all_entities
...
+      options = { inactivity_timeout: 0, keepalive: true }
+
+      EventMachine.run do
+        multi_request = EventMachine::MultiRequest.new
+
+        entities.each do |entity, version|
```
Comment (Fryguy) [inline lib/kubeclient.rb]: So is each entity a separate connection?

File: lib/kubeclient.rb [hunk 4]
```diff
@@ -138,5 +129,51 @@ def all_entities
...
+
+          unless version.nil?
+            uri.query = URI.encode_www_form('resourceVersion' => version)
+          end
+
+          request = EventMachine::HttpRequest.new(uri, options).get
+
+          request.errback do
+            EventMachine.stop
+            fail request.error
```
Comment (Fryguy) [inline lib/kubeclient.rb]: Prefer raise over fail

File: lib/kubeclient.rb [hunk 5]
```diff
@@ -138,5 +129,51 @@ def all_entities
...
+                request.response_header.status,
+                request.response_header.http_reason
+              )
+            end
+          end
+
+          buffer = ''
+          request.stream do |chunk|
+            if request.response_header.successful?
+              buffer += chunk
```
Comment (Fryguy) [inline lib/kubeclient.rb]: Prefer << over += for string concatenation

File: lib/kubeclient.rb [hunk 6]
```diff
@@ -138,5 +129,51 @@ def all_entities
...
+              while (notice = buffer.slice!(/.+\n/))
```
Comment (Fryguy) [inline lib/kubeclient.rb]: If multiple messages come through,  this might miss one,  I think, since the regex is greedy

File: test/watch_test.rb [hunk 7]
```diff
@@ -13,28 +13,36 @@ def test_watch_pod_success
...
                  status: 200)
 
     client = Kubeclient::Client.new 'http://localhost:8080/api/', 'v1beta3'
-    client.watch_pods.with_index do |message, index|
-      assert_instance_of(WatchNotice, message)
-      assert_equal(expected[index]['type'], message.type)
-      assert_equal('Pod', message.object.kind)
-      assert_equal('php', message.object.metadata.name)
+
+    watch_enum = Enumerator.new do |x, y|
```
Comment (Fryguy) [inline test/watch_test.rb]: I don't understand using enumerator here... Why not a normal array?

**Comment by Fryguy** [comment]
Doh I just noticed you closed this.   It seemed small enough for a PR to me so not sure why you closed it.

**Comment by simon3z** [comment]
@Fryguy the problem is that em-http-request has terrible ssl support. At the moment ssl support is more important than multi-watch. I'm evaluating typhoeus as an alternative.

**Pull request closed without merging**